twenty-first/src/util_types/mmr/mmr_trait.rs

```diff
@@ -39,42 +39,129 @@ impl LeafMutation {
     }
 }
 
+/// A “Merkle Mountain Range” (MMR) is a data structure for storing lists of
+/// hashes. Similar to [Merkle trees](crate::prelude::MerkleTree), they enable
+/// efficient set inclusion proofs. Unlike Merkle trees, the number of leafs of
+/// an MMR is not required to be a power of 2.
+///
+/// One valid perspective of this data structure is to see it as a list of
+/// Merkle trees. Each of the Merkle tree's roots is one peak of the MMR.
+/// In order to keep the number of peaks to a minimum, any two neighboring
+/// Merkle trees of the same height are merged, which results in a single tree
+/// that has one additional “tier”.
+///
+/// # Example
+///
+/// The following is a Merkle Mountain Range, annotated with both node and leaf
+/// indices:
+///
+/// ```markdown
+///         ──── 14 ───                           ╮
+///        ╱           ╲                          │
+///       6             13            21          │
+///      ╱  ╲          ╱  ╲          ╱  ╲         ├╴ node indices
+///     ╱    ╲        ╱    ╲        ╱    ╲        │
+///    2      5      9      12     17     20      │
+///   ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲      │
+///  0   1  3   4  7   8  10 11  15 16  18 19 22  ╯
+///
+///  0   1  2   3  4   5  6   7  8   9  10 11 12  ←── leaf indices
+/// ```
+///
+/// Adding another leaf to the above MMR will result in the following MMR:
+///
+/// ```markdown
+///         ──── 14 ───
+///        ╱           ╲
+///       6             13            21
+///      ╱  ╲          ╱  ╲          ╱  ╲
+///     ╱    ╲        ╱    ╲        ╱    ╲
+///    2      5      9      12     17     20    24
+///   ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲   ╱ ╲
+///  0   1  3   4  7   8  10 11  15 16  18 19 22 23
+///
+///  0   1  2   3  4   5  6   7  8   9  10 11 12 13
+/// ```
+///
+/// Note how the new leaf with index 13, which by itself is a Merkle tree of
+/// height 0, was merged with its neighboring Merkle tree of equal height,
+/// resulting in a new Merkle tree of height 1 (the nodes 22, 23, and 24).
+///
+/// Adding two more leafs to this MMR results in a single Merkle tree of height
+/// four.
+///
+/// # Counterexample
+///
+/// The following is **not** a valid Merkle Mountain Range:
+///
+/// ```markdown
+///       _                        _
+///      ╱  ╲                     ╱  ╲
+///     ╱    ╲                   ╱    ╲
+///    _      _     _     _     _      _
+///   ╱ ╲    ╱ ╲   ╱ ╲   ╱ ╲   ╱ ╲    ╱ ╲
+///  _   _  _   _ _   _ _   _ _   _  _   _
+/// ```
+///
+/// It has multiple problems: There are two neighboring Merkle trees of equal
```

Comment:
It's enough of a violation that there are two Merkle trees of equal height, they don't have to be neighboring.